tmt/cli.py

```diff
@@ -1171,7 +1174,7 @@ def stories_id(
     are kept intact.
     """
     tmt.Story._save_context(context)
-    for story in context.obj.tree.stories():
+    for story in cast(tmt.Tree, context.obj.tree).stories():
```

Comment:
Actually, there are not enough `cast()` calls, as I just realized:

```python
def plans_show(context: click.core.Context, **kwargs: Any) -> None:
    ...
    tmt.Plan._save_context(context)
    for plan in context.obj.tree.plans():
        plan.show()
        echo()
```

`context.obj` is annotated as `Any`, mypy cannot infer it's actually `tmt.cli.ContextObject`. This means, wherever `context.obj` is accessed, we should cast it to `ContextObject` so mypy could catch up and follow the trail:

```python
def plans_show(context: click.core.Context, **kwargs: Any) -> None:
    ...
    tmt.Plan._save_context(context)
    for plan in cast(ContextObject, context.obj).tree.plans():
        plan.show()
        echo()
```

So, I should review `cli.py` and add these calls. I suppose that would make you raise your eyebrows even more :) Another idea: how about adding a custom subclass of `click.core.Context` which would only serve as a type stub, e.g.:

```python
class Context(click.core.Context):
  obj: ContextObject  # yeah, maybe we should also rename this class, there's way too many contexts and objects...

...

# Then this should work:
def plans_show(context: Context, **kwargs: Any) -> None:
    ...
    assert context.obj is not None
    tmt.Plan._save_context(context)
    for plan in context.obj.tree.plans():
        plan.show()
        echo()
```

No more `cast()`, but we should replace all `click.Context` annotations in code with the custom class, to make things right (all commands that take `context` parameter, for example...). This plan would actually be my first option, sounds like the least expensive introduction of the "context.obj is of `ContextObject` type" concept. verbose, absolutely, but also fairly correct.

What do you think?

Reply:
I have a patch with a custom `Context` subclass ready, just in case :) It's 857 lines long, but oh boy, what a safety! :) 5 `cast()` calls were removed.

Reply:
Ok, let's rather keep the patch for another pull request, not to block this one ;-)

Reply:
Ok, submitted as an extra PR, https://github.com/teemtee/tmt/pull/1592.